redisvl/llmcache/semantic.py

```diff
@@ -0,0 +1,126 @@
+
+from redisvl.index import SearchIndex
+from redisvl.llmcache.base import BaseLLMCache
+from redisvl.providers import HuggingfaceProvider
+from redisvl.providers.base import BaseProvider
+from redisvl.query import create_vector_query
+from redisvl.utils.log import get_logger
+from redisvl.utils.utils import array_to_buffer
+
+_logger = get_logger(__name__)
+
+
+class SemanticCache(BaseLLMCache):
+    """Cache for Large Language Models."""
+
+    _default_fields = [
+        VectorField(
+            "prompt_vector",
+            "FLAT",
+            {"DIM": 768, "TYPE": "FLOAT32", "DISTANCE_METRIC": "COSINE"},
+        ),
+    ]
+    _default_provider = HuggingfaceProvider("sentence-transformers/all-mpnet-base-v2")
+
+    def __init__(
+        self,
+        index_name: str = "cache",
+        prefix: str = "llmcache",
+        threshold: float = 0.9,
+        provider: Optional[BaseProvider] = None,
+        redis_url: Optional[str] = "redis://localhost:6379",
+        connection_args: Optional[dict] = None,
+        ttl: Optional[int] = None,
+    ):
+
+        self.ttl = ttl
+        self._provider = provider or self._default_provider
+        self._threshold = threshold
+
+        # TODO - configure logging based on verbosity
+        self._index = SearchIndex(
+            index_name, prefix=prefix, fields=self._default_fields
+        )
+        connection_args = connection_args or {}
+        self._index.connect(redis_url=redis_url, **connection_args)
+        self._index.create()
+
+    @property
+    def index(self) -> SearchIndex:
+        """Returns the index for the cache."""
+        return self._index
+
+    @property
+    def threshold(self) -> float:
+        """Returns the threshold for the cache."""
+        return self._threshold
+
+    def set_threshold(self, threshold: float):
+        """Sets the threshold for the cache."""
+        self._threshold = threshold
+
+    def check(
+        self,
+        prompt: Optional[str] = None,
+        vector: Optional[List[float]] = None,
+        num_results: int = 1,
+        fields: List[str] = ["response"],
+    ) -> Optional[List[str]]:
+        """Checks whether the cache contains the specified key."""
+        if not prompt and not vector:
+            raise ValueError("Either prompt or vector must be specified.")
+
+        query = create_vector_query(
+            return_fields=fields,
+            vector_field_name="prompt_vector",
+            number_of_results=num_results,
+        )
+        if vector:
+            prompt_vector = array_to_buffer(vector)
+        else:
+            prompt_vector = array_to_buffer(self._provider.embed(prompt)) # type: ignore
+        results = self._index.search(query, query_params={"vector": prompt_vector})
+
+        cache_hits = []
+        for doc in results.docs:
+            sim = similarity(doc.vector_score)
+            _logger.info("Similarity: %s", sim)
+            if similarity(doc.vector_score) > self.threshold:
```

Comment:
we should take advantage of a range query for this. more performant and built into the search API. I think all we need is to expose a different `create_range_query` method in addition to `create_knn_query`* maybe?

Reply:
Let's chat about this. I agree, just want to make sure we are thinking of the same thing. This would be a good one for you to pick up on.